src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+/// and will continue to update v1.
+///
+/// This maintains a filesystem lock, preventing other instances of Cargo from
+/// modifying at the same time. Drop the value to unlock.
+pub struct InstallTracker {
+    v1: CrateListingV1,
+    v2: CrateListingV2,
+    v1_lock: FileLock,
+    v2_lock: Option<FileLock>,
+    unstable_upgrade: bool,
+}
+
+#[derive(Default, Deserialize, Serialize)]
+struct CrateListingV2 {
+    installs: BTreeMap<PackageId, InstallInfo>,
+    /// Forwards compatibility.
+    #[serde(flatten)]
+    other: BTreeMap<String, serde_json::Value>,
 }
 
-#[derive(Deserialize, Serialize)]
-#[serde(deny_unknown_fields)]
-pub struct Empty {}
+#[derive(Debug, Deserialize, Serialize)]
+struct InstallInfo {
+    /// Version requested via `--version`.
+    /// None if `--version` not specified. Currently not used, possibly may be
+    /// used in the future.
+    version_req: Option<String>,
+    /// Set of binary names installed.
+    bins: BTreeSet<String>,
+    /// Set of features explicitly enabled.
+    features: BTreeSet<String>,
+    all_features: bool,
+    no_default_features: bool,
+    /// Either "debug" or "release".
+    profile: String,
+    /// The installation target.
+    /// Either the host or the value specified in `--target`.
+    /// None if unknown (when loading from v1).
+    target: Option<String>,
+    /// Output of `rustc -V`.
+    /// None if unknown (when loading from v1).
+    /// Currently not used, possibly may be used in the future.
+    rustc: Option<String>,
+    /// Forwards compatibility.
+    #[serde(flatten)]
+    other: BTreeMap<String, serde_json::Value>,
+}
 
-#[derive(Deserialize, Serialize)]
+#[derive(Default, Deserialize, Serialize)]
 pub struct CrateListingV1 {
     v1: BTreeMap<PackageId, BTreeSet<String>>,
 }
 
+impl InstallTracker {
+    /// Create an InstallTracker from information on disk.
+    pub fn load(config: &Config, root: &Filesystem) -> CargoResult<InstallTracker> {
+        let unstable_upgrade = config.cli_unstable().install_upgrade;
+        let v1_lock = root.open_rw(Path::new(".crates.toml"), config, "crate metadata")?;
+        let v2_lock = if unstable_upgrade {
+            Some(root.open_rw(Path::new(".crates2.json"), config, "crate metadata")?)
+        } else {
+            None
+        };
+
+        let v1 = (|| -> CargoResult<_> {
+            let mut contents = String::new();
+            v1_lock.file().read_to_string(&mut contents)?;
+            if contents.is_empty() {
+                Ok(CrateListingV1::default())
+            } else {
+                Ok(toml::from_str(&contents)
+                    .chain_err(|| format_err!("invalid TOML found for metadata"))?)
+            }
+        })()
+        .chain_err(|| {
+            format_err!(
+                "failed to parse crate metadata at `{}`",
+                v1_lock.path().to_string_lossy()
+            )
+        })?;
+
+        let v2 = (|| -> CargoResult<_> {
+            if unstable_upgrade {
```

Comment:
Since `unstable_upgrade` is already checked abvoe I think this can just be a `match` on `v2_lock`?